src/wallet/wallet.cpp

```diff
@@ -1403,20 +1403,30 @@ bool CWallet::IsAllFromMe(const CTransaction& tx, const isminefilter& filter) co
 {
     LOCK(cs_wallet);
 
-    for (const CTxIn& txin : tx.vin)
-    {
+    std::vector<const CTxOut*> outs;
+    outs.reserve(tx.vin.size());
+
+    // first check if inputs are known and valid
+    for (const CTxIn& txin : tx.vin) {
         auto mi = mapWallet.find(txin.prevout.hash);
-        if (mi == mapWallet.end())
-            return false; // any unknown inputs can't be from us
 
-        const CWalletTx& prev = (*mi).second;
+        // any unknown inputs can't be from us
+        if (mi == mapWallet.end()) return false;
 
-        if (txin.prevout.n >= prev.tx->vout.size())
-            return false; // invalid input!
+        const CWalletTx& prev = mi->second;
+        // check for invalid input
+        if (txin.prevout.n >= prev.tx->vout.size()) return false;
 
-        if (!(IsMine(prev.tx->vout[txin.prevout.n]) & filter))
-            return false;
+        outs.push_back(&prev.tx->vout[txin.prevout.n]);
     }
+
+    assert(outs.size() == tx.vin.size());
+
+    // then check if every isminetype matches filter
```

Comment:
Would suggest adding a comment that the code is intentionally written to use 2 passes for better performance. Otherwise it's unclear why the code would be organized this way.

Reply:
Done. PR title also mentions the idea behind the 2 loops.